streams-tests/jvm/src/test/scala/zio/stream/SinkSpec.scala

```diff
@@ -17,29 +17,791 @@ class SinkSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
+    def extract(state: State)      = IO.fail("Ouch")
+  }
+
+  private def stepErrorSink = new ZSink[Any, String, Int, Int, Int] {
+    type State = Unit
+    val initial                    = UIO.succeed(Step.more(()))
+    def step(state: State, a: Int) = IO.fail("Ouch")
+    def extract(state: State)      = IO.fail("Ouch")
+  }
+
+  private def extractErrorSink = new ZSink[Any, String, Int, Int, Int] {
+    type State = Unit
+    val initial                    = UIO.succeed(Step.more(()))
+    def step(state: State, a: Int) = UIO.succeed(Step.done((), Chunk.empty))
+    def extract(state: State)      = IO.fail("Ouch")
+  }
+
+  private def sinkIteration[R, E, A0, A, B](sink: ZSink[R, E, A0, A, B], a: A) =
+    for {
+      init   <- sink.initial
+      step   <- sink.step(Step.state(init), a)
+      result <- sink.extract(Step.state(step))
+    } yield result
+
+  private def chunkedHappyPath = {
+    val sink = ZSink.collectAll[Int].chunked
+    unsafeRun(sinkIteration(sink, Chunk(1, 2, 3, 4, 5)).map(_ must_=== List(1, 2, 3, 4, 5)))
+  }
+
+  private def chunkedEmpty = {
+    val sink = ZSink.collectAll[Int].chunked
+    unsafeRun(sinkIteration(sink, Chunk.empty).map(_ must_=== Nil))
+  }
+
+  private def chunkedInitError = {
+    val sink = initErrorSink.chunked
+    unsafeRun(sinkIteration(sink, Chunk.single(1)).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def chunkedStepError = {
+    val sink = stepErrorSink.chunked
+    unsafeRun(sinkIteration(sink, Chunk.single(1)).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def chunkedExtractError = {
+    val sink = extractErrorSink.chunked
+    unsafeRun(sinkIteration(sink, Chunk.single(1)).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def collectAllHappyPath = {
+    val sink = ZSink.identity[Int].collectAll[Int, Int]
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== List(1)))
+  }
+
+  private def collectAllInitError = {
+    val sink = initErrorSink.collectAll
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def collectAllStepError = {
+    // This test needs to be verified for correctness.
+    val sink = stepErrorSink.collectAll
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== Nil))
+  }
+
+  private def collectAllExtractError = {
+    val sink = extractErrorSink.collectAll
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def collectAllWhileHappyPath = {
+    val sink = ZSink.identity[Int].collectAllWhile[Int, Int](_ < 10)
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== List(1)))
+  }
+
+  private def collectAllWhileFalsePredicate = {
+    // This test needs to be verified for correctness.
+    // I find this behavior to be surprising.
+    val sink = ZSink.identity[Int].collectAllWhile[Int, Int](_ < 0)
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left(())))
+    // Fails instead of returning empty list.
```

Comment:
The implementation of `collectAllWhile` looks right - it tries to extract the underlying sink if an element that doesn't match the predicate is stepped.

The problem is with `ZSink.identity`, however; that one fails with `Unit` if an element never passed through the sink. Maybe test it with a different sink?